### ISSUE TITLE
ConcurrentDictionary.Clear ignores the capacity specified in the constructor

### ISSUE THREAD

**Issue by MTBandolier** [dotnet/corefx#26899]
The Clear method always allocates DEFAULT_CAPACITY Nodes instead of using the capacity specified in the constructor.  It should probably use m_tables.m_buckets.Length instead.

```cs
public void Clear()
{
    int locksAcquired = 0;
    try
    {
        AcquireAllLocks(ref locksAcquired);

        Tables newTables = new Tables(new Node[DEFAULT_CAPACITY], m_tables.m_locks, new int[m_tables.m_countPerLock.Length], m_tables.m_comparer);
        m_tables = newTables;
        m_budget = Math.Max(1, newTables.m_buckets.Length / newTables.m_locks.Length);
    }
    finally
    {
        ReleaseLocks(0, locksAcquired);
    }
}
```

**Comment by danmosemsft** [comment]
I agree it does not seem right that Clear() will shrink the storage. Clear() of a regular Dictionary does not. Unless @stephentoub there is some reason? I cannot find it in history.

**Comment by stephentoub** [comment]
This is a very different situation and implementation.  On Dictionary, Clear doesn't shrink storage; it keeps the same array and updates some bookkeeping.  On ConcurrentDictionary, however, Clear has to deal with concurrent readers, and thus implements the operation by replacing the existing tables with a new one, which means allocating a new buckets array that's swapped in, leaving the old one intact for any readers still in flight.  To not "shrink the storage", ConcurrentDictionary would have to allocate a potentially gigantic array (if the previous buckets had grown to be large), making it much more expensive to not shrink the storage.  Further, CD currently doesn't even store the original capacity (m_tables.m_buckets.Length is not the capacity stored in the constructor, but rather what the "capacity" has grown to)... I also put "capacity" in quotes because that table isn't actually storing the data directly, but rather is storing a linked list of node objects, which are necessarily going to be thrown away as part of the Clear.

**Comment by danmosemsft** [comment]
Makes sense. @MTBandolier is this just an observation or are you affected?

**Comment by Duranom** [comment]
Should the reasoning of @stephentoub be be put in code comment?

**Comment by LeaFrock** [comment]
I agree with stephentoub's answer above. But maybe we hope _Clear()_ of CD keeps its capacity when we can anticipate that there'll be a lot of add-operations soon. So I have a suggestion: how about add a overload-method of _Clear()_ like _Clear(int capacity)_？Give programmers a flexible choice.
